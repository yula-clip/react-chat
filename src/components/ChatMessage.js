import React from 'react';
import { withStyles } from '@material-ui/core/styles';
import classnames from 'classnames';
import moment from 'moment';
import Typography from '@material-ui/core/Typography';
import Paper from '@material-ui/core/Paper';
import Avatar from './Avatar';

import senderName from '../utils/sender-name';
import randomColor from '../utils/color-from';

const styles = theme =>({
  messageWrapperFromMe:{
    justifyContent: 'flex-end !important',
  },
  message: {
    maxWidth: '70%',
    minWidth: '10%',
    padding: theme.spacing.unit,
    marginLeft: theme.spacing.unit * 2,
  },
  messageFromMe: {
    marginRight: theme.spacing.unit * 2,
    backgroundColor: '#e6dcff',
  },
  messageWrapper:{
    display: 'flex',
    justifyContent: 'flex-start',
    alignItems: 'center',
    padding: `${theme.spacing.unit}px ${theme.spacing.unit * 3}px`,
  },
  statusMessage: {
    width: '100%',
    textAlign: 'center',
  },
  statusMessageUser: {
    display: 'inline',
  },
});

const ChatMessage = ({ classes, sender, content, activeUser, createdAt, statusMessage }) => {
  const isMessageFromMe = sender._id === activeUser._id;

  const displayedName = senderName(sender);

  if (statusMessage) {
    return (
      <div className={classes.messageWrapper}>
        <Typography className={classes.statusMessage}>
          <Typography 
            variant="caption" 
            style={{ color: randomColor(sender._id)}} 
            className={classes.statusMessageUser}
          >
            {displayedName}
          </Typography>
          {content}
          <Typography variant="caption" component="span">
            {moment(createdAt).fromNow()}
          </Typography>
        </Typography>
      </div>
    )
  }

  const userAvatar = (
    <Avatar colorFrom = {sender._id}>
      {displayedName}
    </Avatar>
  );

  return (
    <div className={classnames(
      classes.messageWrapper,
      isMessageFromMe && classes.messageWrapperFromMe
    )}>
      {!isMessageFromMe && userAvatar}
      <Paper className={classnames(
        classes.message,
        isMessageFromMe && classes.messageFromMe
      )}>
        <Typography variant="caption" style={{ color: randomColor(sender._id)}}>
          {displayedName}
        </Typography>
        <Typography variant="body1">
          {content}
        </Typography>
        <Typography variant="caption" className={classes.time}>
          {moment(createdAt).fromNow()}
        </Typography>
      </Paper>
      {isMessageFromMe && userAvatar}
    </div>
  );
};

export default withStyles(styles)(ChatMessage);
