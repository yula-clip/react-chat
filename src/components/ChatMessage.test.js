/* eslint-env jest */
import React from "react";
import ReactDOM from "react-dom";
import renderer from "react-test-renderer";
import ChatMessage from "./ChatMessage";

jest.mock("moment", () => () => ({ fromNow: () => "2 days ago" }));

const mockProps = {
  content: "Hello, World!",
  sender: {
    _id: "12345",
    username: "me"
  },
  activeUser: {
    _id: "12345"
  },
  createdAt: "2018-03-16T10:53:23.200Z",
  statusMessage: false
};

describe("<ChatMessage />", () => {
  it("renders without crashing", () => {
    const div = document.createElement("div");
    ReactDOM.render(<ChatMessage {...mockProps} />, div);
    ReactDOM.unmountComponentAtNode(div);
  });

  it("renders message from active user", () => {
    const tree = renderer.create(<ChatMessage {...mockProps} />).toJSON();

    expect(tree).toMatchSnapshot();
  });

  it("renders name of active user correctly", () => {
    const tree = renderer
      .create(
        <ChatMessage
          {...mockProps}
          sender={{
            _id: "12345",
            username: "me",
            firstName: "Name",
            lastName: "Surname"
          }}
        />
      )
      .toJSON();

    expect(tree).toMatchSnapshot();
  });

  it("renders message from other users", () => {
    const tree = renderer
      .create(
        <ChatMessage
          {...mockProps}
          sender={{
            _id: "54321",
            username: "someone"
          }}
        />
      )
      .toJSON();

    expect(tree).toMatchSnapshot();
  });

  it("renders name of other users correctly", () => {
    const tree = renderer
      .create(
        <ChatMessage
          {...mockProps}
          sender={{
            _id: "54321",
            username: "someone",
            firstName: "Name",
            lastName: "Surname"
          }}
        />
      )
      .toJSON();

    expect(tree).toMatchSnapshot();
  });

  it("renders status message", () => {
    const tree = renderer
      .create(<ChatMessage {...mockProps} statusMessage />)
      .toJSON();

    expect(tree).toMatchSnapshot();
  });

  it("renders name in status message correctly", () => {
    const tree = renderer
      .create(
        <ChatMessage
          {...mockProps}
          sender={{
            _id: "54321",
            username: "someone",
            firstName: "Name",
            lastName: "Surname"
          }}
          statusMessage
        />
      )
      .toJSON();

    expect(tree).toMatchSnapshot();
  });
});
