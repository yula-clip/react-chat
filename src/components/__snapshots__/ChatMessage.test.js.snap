// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ChatMessage /> renders message from active user 1`] = `
<div
  className="ChatMessage-messageWrapper-75 ChatMessage-messageWrapperFromMe-72"
>
  <div
    className="MuiPaper-root-78 MuiPaper-elevation2-82 MuiPaper-rounded-79 ChatMessage-message-73 ChatMessage-messageFromMe-74"
  >
    <span
      className="MuiTypography-root-105 MuiTypography-caption-115"
      style={
        Object {
          "color": "#f44336",
        }
      }
    >
      me
    </span>
    <p
      className="MuiTypography-root-105 MuiTypography-body1-114"
    >
      Hello, World!
    </p>
    <span
      className="MuiTypography-root-105 MuiTypography-caption-115"
    >
      2 days ago
    </span>
  </div>
  <div
    className="MuiAvatar-root-140 MuiAvatar-colorDefault-141"
    style={
      Object {
        "backgroundColor": "#f44336",
      }
    }
  >
    M
  </div>
</div>
`;

exports[`<ChatMessage /> renders message from other users 1`] = `
<div
  className="ChatMessage-messageWrapper-75"
>
  <div
    className="MuiAvatar-root-140 MuiAvatar-colorDefault-141"
    style={
      Object {
        "backgroundColor": "#f44336",
      }
    }
  >
    S
  </div>
  <div
    className="MuiPaper-root-78 MuiPaper-elevation2-82 MuiPaper-rounded-79 ChatMessage-message-73"
  >
    <span
      className="MuiTypography-root-105 MuiTypography-caption-115"
      style={
        Object {
          "color": "#f44336",
        }
      }
    >
      someone
    </span>
    <p
      className="MuiTypography-root-105 MuiTypography-body1-114"
    >
      Hello, World!
    </p>
    <span
      className="MuiTypography-root-105 MuiTypography-caption-115"
    >
      2 days ago
    </span>
  </div>
</div>
`;

exports[`<ChatMessage /> renders name in status message correctly 1`] = `
<div
  className="ChatMessage-messageWrapper-75"
>
  <p
    className="MuiTypography-root-105 MuiTypography-body1-114 ChatMessage-statusMessage-76"
  >
    <span
      className="MuiTypography-root-105 MuiTypography-caption-115 ChatMessage-statusMessageUser-77"
      style={
        Object {
          "color": "#f44336",
        }
      }
    >
      Name Surname
    </span>
    Hello, World!
    <span
      className="MuiTypography-root-105 MuiTypography-caption-115"
    >
      2 days ago
    </span>
  </p>
</div>
`;

exports[`<ChatMessage /> renders name of active user correctly 1`] = `
<div
  className="ChatMessage-messageWrapper-75 ChatMessage-messageWrapperFromMe-72"
>
  <div
    className="MuiPaper-root-78 MuiPaper-elevation2-82 MuiPaper-rounded-79 ChatMessage-message-73 ChatMessage-messageFromMe-74"
  >
    <span
      className="MuiTypography-root-105 MuiTypography-caption-115"
      style={
        Object {
          "color": "#f44336",
        }
      }
    >
      Name Surname
    </span>
    <p
      className="MuiTypography-root-105 MuiTypography-body1-114"
    >
      Hello, World!
    </p>
    <span
      className="MuiTypography-root-105 MuiTypography-caption-115"
    >
      2 days ago
    </span>
  </div>
  <div
    className="MuiAvatar-root-140 MuiAvatar-colorDefault-141"
    style={
      Object {
        "backgroundColor": "#f44336",
      }
    }
  >
    NS
  </div>
</div>
`;

exports[`<ChatMessage /> renders name of other users correctly 1`] = `
<div
  className="ChatMessage-messageWrapper-75"
>
  <div
    className="MuiAvatar-root-140 MuiAvatar-colorDefault-141"
    style={
      Object {
        "backgroundColor": "#f44336",
      }
    }
  >
    NS
  </div>
  <div
    className="MuiPaper-root-78 MuiPaper-elevation2-82 MuiPaper-rounded-79 ChatMessage-message-73"
  >
    <span
      className="MuiTypography-root-105 MuiTypography-caption-115"
      style={
        Object {
          "color": "#f44336",
        }
      }
    >
      Name Surname
    </span>
    <p
      className="MuiTypography-root-105 MuiTypography-body1-114"
    >
      Hello, World!
    </p>
    <span
      className="MuiTypography-root-105 MuiTypography-caption-115"
    >
      2 days ago
    </span>
  </div>
</div>
`;

exports[`<ChatMessage /> renders status message 1`] = `
<div
  className="ChatMessage-messageWrapper-75"
>
  <p
    className="MuiTypography-root-105 MuiTypography-body1-114 ChatMessage-statusMessage-76"
  >
    <span
      className="MuiTypography-root-105 MuiTypography-caption-115 ChatMessage-statusMessageUser-77"
      style={
        Object {
          "color": "#f44336",
        }
      }
    >
      me
    </span>
    Hello, World!
    <span
      className="MuiTypography-root-105 MuiTypography-caption-115"
    >
      2 days ago
    </span>
  </p>
</div>
`;
